vllm/model_executor/layers/quantization/deepspeedfp.py

```diff
@@ -0,0 +1,189 @@
+        return [
+            "quant_config.json",
+            "quantize_config.json",
+        ]
+
+    def get_quant_method(
+            self,
+            layer: torch.nn.Module) -> Optional["DeepSpeedFPLinearMethod"]:
+        if isinstance(layer, LinearBase):
+            return DeepSpeedFPLinearMethod(self)
+        return None
+
+
+class DeepSpeedFPLinearMethod(LinearMethodBase):
+    """Linear method for DeepSpeedFP quantizer.
+
+    Args:
+        quant_config: the DeepSpeedFP quantization config.
+    """
+
+    def __init__(self, quant_config: DeepSpeedFPConfig):
+        self.quant_config = quant_config
+        self.weight = None
+
+    def create_weights(self,
+                       layer: torch.nn.Module,
+                       input_size_per_partition: int,
+                       output_partition_sizes: List[int],
+                       input_size: int,
+                       output_size: int,
+                       params_dtype: torch.dtype,
+                       weight_loader=None,
+                       **extra_weight_attrs):
+        del output_size
+        del input_size
+        output_size_per_partition = sum(output_partition_sizes)
+        weight = DeepSpeedFPParameter(
+            torch.Size((output_size_per_partition, input_size_per_partition)),
+            params_dtype=params_dtype,
+            quant_config=self.quant_config,
+        )
+        set_weight_attrs(weight, {
+            "input_dim": 1,
+            "output_dim": 0,
+        })
+        layer.register_parameter("weight", weight)
+
+        def quant_weight_loader(param, loaded_weight, *args, **kwargs):
+            # Calls the original weight loader (if any), quantizes the result,
+            # and then loads the quantized parameter.
+            if weight_loader is not None:
+                orig_param_data = param.data
+                param.data = param.ds_dequantize()
+                weight_loader(param, loaded_weight, *args, **kwargs)
+                param.data, loaded_weight = orig_param_data, param.data
+            param.ds_quantize_(loaded_weight.cuda())
+
+        extra_weight_attrs["weight_loader"] = quant_weight_loader
+        set_weight_attrs(weight, extra_weight_attrs)
+
+    def apply(self,
+              layer: torch.nn.Module,
+              x: torch.Tensor,
+              bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        weight = layer.weight
+        y = weight.ds_dequantize()
+        return F.linear(x, y, bias)
+
+
+class DeepSpeedFPParameter(nn.Parameter):
+    """
+    DeepSpeedFP quantized parameter class that implements fp8/fp6
+    quantization deepspeed. Weights are stored in quantized form on
+    GPUs, and can be dequantized on-the-fly when needed by the model.
+    """
+
+    def __new__(cls, orig_shape: torch.Size, params_dtype: torch.dtype,
+                quant_config: DeepSpeedFPConfig):
+        from deepspeed.ops.fp_quantizer import FP_Quantize
```

Comment:
Can we add a comment to note the FP_Quantize version we've tested against to? And it would be better to throw an error with a proper error message when ImportError.

Reply:
`FP_Quantize` requires `deepspeed>=0.14.2`. Is there a current way to indicate this type of dependency in vllm?

Reply:
This is now a soft/lazy dependency so we don't put it in `requirements.txt`. What I meant was just add a checker here like

```python
try:
  from deepspeed.ops.fp_quantizer import FP_Quantize
  if FP_Quantize.__version__ < 0.14.2:
    raise ImportError("wrong version")
except ImportError as err:
    raise ImportError(f"nicer message from {err}") from None
```

Reply:
Ah yeah, no worries. The reason I asked is `transformers` has a few nice utils that are used throughout their codebase for optional/lazy dependencies like this (examples below). If there's not then would be nice to use something like `packaging` for the version comparison if possible. 

* https://github.com/huggingface/transformers/blob/5b7a225f2502840f4066b1b39ef53411ea4d6e90/src/transformers/utils/versions.py#L49
* https://github.com/huggingface/transformers/blob/5b7a225f2502840f4066b1b39ef53411ea4d6e90/src/transformers/integrations/deepspeed.py#L34-L47

Reply:
I don't have much content on this. Maybe @simon-mo can comment more.

Reply:
fixed.